### DISCUSSION TITLE
FastAPI Web App with authentication

### DISCUSSION THREAD

**Discussion by GG6666** [fastapi/fastapi#11109] [Questions]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
class OAuth2PasswordBearerWithCookie(OAuth2):
    """
    This class is taken directly from FastAPI:
    https://github.com/tiangolo/fastapi/blob/26f725d259c5dbe3654f221e608b14412c6b40da/fastapi/security/oauth2.py#L140-L171
    
    The only change made is that authentication is taken from a cookie
    instead of from the header!
    """
    def __init__(
        self,
        tokenUrl: str,
        scheme_name: Optional[str] = None,
        scopes: Optional[Dict[str, str]] = None,
        description: Optional[str] = None,
        auto_error: bool = True,
    ):
        if not scopes:
            scopes = {}
        flows = OAuthFlowsModel(password={"tokenUrl": tokenUrl, "scopes": scopes})
        super().__init__(
            flows=flows,
            scheme_name=scheme_name,
            description=description,
            auto_error=auto_error,
        )

    async def __call__(self, request: Request) -> Optional[str]:
        # IMPORTANT: this is the line that differs from FastAPI. Here we use 
        # `request.cookies.get(settings.COOKIE_NAME)` instead of 
        # `request.headers.get("Authorization")`
        authorization: str = request.cookies.get(settings.COOKIE_NAME) 
        scheme, param = get_authorization_scheme_param(authorization)
        if not authorization or scheme.lower() != "bearer":
            if self.auto_error:
                raise HTTPException(
                    status_code=status.HTTP_401_UNAUTHORIZED,
                    detail="Not authenticated",
                    headers={"WWW-Authenticate": "Bearer"},
                )
            else:
                return None
        return param
```

### Description

I found Sam's blog post [https://samedwardes.com/2022/04/14/fastapi-webapp-with-auth/](https://samedwardes.com/2022/04/14/fastapi-webapp-with-auth/). It sounds he had difficulty to figure out how to store JWT tokens in the HTTP request/response header. He worked around to store JWT tokens in the cookie.  --> **FastAPI comes with a class out of the box named [OAuth2PasswordBearer](https://github.com/tiangolo/fastapi/blob/26f725d259c5dbe3654f221e608b14412c6b40da/fastapi/security/oauth2.py#L140-L171) (see FastAPI docs: [Security - First Steps](https://fastapi.tiangolo.com/tutorial/security/first-steps/?h=oauth2passwordbearer#fastapis-oauth2passwordbearer)). This is useful for authenticating APIs. However, I could not figure out how to use it in the context of a web app.** Has somebody figured out how to authenticate FastAPI web application by storing/retrieving JWT tokens in HTTP request/response header?

### Operating System

Linux

### FastAPI Version

0.109.2

### Pydantic Version

2.6.1

### Python Version

3.11.6

**Comment by jgould22** [comment]
JWTs can be used as bearer tokens.

You can probably use the HTTPBearer and put your jwt in the `Authorization` header

https://github.com/tiangolo/fastapi/blob/master/fastapi/security/http.py#L220

Then in a dependency sort out whatever you need to with your jwt

**Comment by JonnyBootsNpants** [comment]
This file contains over 2,130 ADS 'alternate data streams' so... yes safety first lol
________________________________
From: GG6666 ***@***.***>
Sent: Wednesday, February 7, 2024 7:36 PM
To: tiangolo/fastapi ***@***.***>
Cc: Subscribed ***@***.***>
Subject: [tiangolo/fastapi] FastAPI Web App with authentication (Discussion #11109)


First Check

  *    I added a very descriptive title here.
  *    I used the GitHub search to find a similar question and didn't find it.
  *    I searched the FastAPI documentation, with the integrated search.
  *    I already searched in Google "How to X in FastAPI" and didn't find any information.
  *    I already read and followed all the tutorial in the docs and didn't find an answer.
  *    I already checked if it is not related to FastAPI but to Pydantic<https://github.com/pydantic/pydantic>.
  *    I already checked if it is not related to FastAPI but to Swagger UI<https://github.com/swagger-api/swagger-ui>.
  *    I already checked if it is not related to FastAPI but to ReDoc<https://github.com/Redocly/redoc>.

Commit to Help

  *    I commit to help with one of those options 👆

Example Code

class OAuth2PasswordBearerWithCookie(OAuth2):
    """
    This class is taken directly from FastAPI:
    https://github.com/tiangolo/fastapi/blob/26f725d259c5dbe3654f221e608b14412c6b40da/fastapi/security/oauth2.py#L140-L171

    The only change made is that authentication is taken from a cookie
    instead of from the header!
    """
    def __init__(
        self,
        tokenUrl: str,
        scheme_name: Optional[str] = None,
        scopes: Optional[Dict[str, str]] = None,
        description: Optional[str] = None,
        auto_error: bool = True,
    ):
        if not scopes:
            scopes = {}
        flows = OAuthFlowsModel(password={"tokenUrl": tokenUrl, "scopes": scopes})
        super().__init__(
            flows=flows,
            scheme_name=scheme_name,
            description=description,
            auto_error=auto_error,
        )

    async def __call__(self, request: Request) -> Optional[str]:
        # IMPORTANT: this is the line that differs from FastAPI. Here we use
        # `request.cookies.get(settings.COOKIE_NAME)` instead of
        # `request.headers.get("Authorization")`
        authorization: str = request.cookies.get(settings.COOKIE_NAME)
        scheme, param = get_authorization_scheme_param(authorization)
        if not authorization or scheme.lower() != "bearer":
            if self.auto_error:
                raise HTTPException(
                    status_code=status.HTTP_401_UNAUTHORIZED,
                    detail="Not authenticated",
                    headers={"WWW-Authenticate": "Bearer"},
                )
            else:
                return None
        return param

Description

I found Sam's blog post https://samedwardes.com/2022/04/14/fastapi-webapp-with-auth/. It sounds he had difficulty to figure out how to store JWT tokens in the HTTP request/response header. He worked around to store JWT tokens in the cookie. --> FastAPI comes with a class out of the box named OAuth2PasswordBearer<https://github.com/tiangolo/fastapi/blob/26f725d259c5dbe3654f221e608b14412c6b40da/fastapi/security/oauth2.py#L140-L171> (see FastAPI docs: Security - First Steps<https://fastapi.tiangolo.com/tutorial/security/first-steps/?h=oauth2passwordbearer#fastapis-oauth2passwordbearer>). This is useful for authenticating APIs. However, I could not figure out how to use it in the context of a web app. Has somebody figured out how to authenticate FastAPI web application by storing/retrieving JWT tokens in HTTP request/response header?

Operating System

Linux

Operating System Details

No response

FastAPI Version

0.109.2

Pydantic Version

2.6.1

Python Version

3.11.6

Additional Context

No response

**Comment by YuriiMotov** [comment]
Please, refer to the docs. I believe, there is enough information there..

Let's close this for now to keep the discussions clean and organized.
If you still need help, feel free to let us know.
